File: Mesh4D/Matlab4D/Tensor/src/OPC4D_Tensor.m
function [XYZT, W] = OPC4D_Tensor(ade, vertici_iniziali, vertici_finali, facets, method)

%**************************************************************************
%
% function [XYZT, W] = OPC4D_Tensor(ade, vertici_iniziali, ...
%                                   vertici_finali, facets)
%
% Costruisce una regola di cubatura quadridimensionale per un dominio
% poliedrico spazio-temporale in (x,y,z,tau), ottenuto dall'evoluzione
% di un dominio poliedrico tridimensionale tra le configurazioni iniziale
% e finale corrispondenti rispettivamente a tau = 0 e tau = 1.
%
% Il dominio 4D è rappresentato mediante una mesh superficiale composta
% da due sezioni tridimensionali, poste agli istanti tau = 0 e tau = 1,
% e dalle corrispondenti iperfacce laterali che collegano le facce
% triangolari della mesh iniziale con quelle della mesh finale.
%
% La regola di cubatura è costruita mediante:
%
%   - una base tensoriale di polinomi di Chebyshev in quattro variabili;
%   - una regola di Gauss-Chebyshev tensoriale sul dominio di riferimento
%     [-1,1]^4;
%   - il calcolo dei momenti shape-dependent della base sul dominio
%     spazio-temporale mediante il Teorema della Divergenza in 4D;
%   - il riscalamento della regola di riferimento sulla bounding box
%     del dominio fisico;
%   - la ricostruzione dei pesi della regola mediante la matrice di
%     Vandermonde di Chebyshev.
%
% La funzione restituisce esclusivamente i nodi e i pesi della regola
% di cubatura 4D. In particolare, per una funzione f(x,y,z,tau), 
% l'integrale può essere calcolato esternamente mediante
%
%       f_val = f(XYZT(:,1), XYZT(:,2), XYZT(:,3), XYZT(:,4));
%       I = sum(W .* f_val);
%
%**************************************************************************
%
% INPUT:
%
%   ade:
%       Grado polinomiale totale massimo della base di Chebyshev
%       utilizzata per la costruzione della regola di cubatura 4D.
%
%       La dimensione dello spazio polinomiale di grado totale <= ade
%       in quattro variabili è
%
%           N = (ade+1)(ade+2)(ade+3)(ade+4) / 24.
%
%   vertici_iniziali:
%       Matrice N_v x 3 contenente le coordinate cartesiane dei vertici
%       della mesh nella configurazione iniziale, corrispondente a
%       tau = 0.
%
%       Ogni riga rappresenta un vertice nella forma
%
%           [x_i, y_i, z_i].
%
%   vertici_finali:
%       Matrice N_v x 3 contenente le coordinate cartesiane dei vertici
%       della mesh nella configurazione finale, corrispondente a
%       tau = 1.
%
%       La connettività della mesh deve essere la stessa utilizzata per
%       vertici_iniziali.
%
%   facets:
%       Matrice N_f x 3 contenente la connettività della mesh superficiale
%       triangolare.
%
%   method:
%       Stringa che specifica la regola di quadratura utilizzata per
%       l'integrazione sulle iperfacce laterali del dominio 4D.
%
%       Sono disponibili i seguenti metodi:
%
%           'DCC'  : Dunavant--Clenshaw--Curtis
%           'DGL'  : Dunavant--Gauss--Legendre
%           'GJCC' : Gauss--Jacobi--Clenshaw--Curtis
%           'GJL'  : Gauss--Jacobi--Gauss--Legendre
%
%       Le regole sono costruite come prodotto tensoriale tra una
%       quadratura sul triangolo di riferimento e una quadratura
%       unidimensionale sull'intervallo temporale [0,1].
%
%**************************************************************************
%
% OUTPUT:
%
%   XYZT:
%       Matrice N_q x 4 contenente i nodi della regola di cubatura
%       quadridimensionale.
%
%       Ogni riga contiene un nodo nella forma
%
%           [x_i, y_i, z_i, tau_i],
%
%       con tau_i appartenente all'intervallo [0,1].
%
%   W:
%       Vettore colonna N_q x 1 contenente i pesi della regola di
%       cubatura quadridimensionale associati ai nodi contenuti in XYZT.
%
%**************************************************************************
%
% NOTE:
%
%   Il dominio spazio-temporale 4D è costruito a partire dalle
%   configurazioni iniziale e finale della mesh. Le iperfacce laterali
%   sono ottenute collegando ciascuna faccia triangolare iniziale con la
%   corrispondente faccia triangolare finale, formando un prisma
%   spazio-temporale tridimensionale.
%
%   Le normali alle iperfacce del bordo 4D vengono orientate verso
%   l'esterno del dominio mediante il confronto con il baricentro
%   globale del dominio spazio-temporale.
%
%**************************************************************************
%
% Autore:
%       Edoardo Longo
%       Università degli Studi di Verona
%
% Data:
%       Settembre 2026
%
%**************************************************************************

% Percorso della directory principale del progetto
projectRoot = fileparts(fileparts(mfilename('fullpath')));

% Aggiunta della cartella contenente le funzioni ausiliarie
addpath(fullfile(projectRoot, 'src'));

%**************************************************************************
%
% INIZIO PARTE SHAPE-INDEPENDENT
%
%**************************************************************************

% Generazione griglia tensoriale di Gauss-Chebyshev sul dominio di riferimento [-1, 1]^4
XYZTW_tens_ref = cub_gausscheb_tens4D(2 * ade);

% Calcolo della dimensione dello spazio polinomiale 4D per il grado 'ade'
N = ((ade + 4) * (ade + 3) * (ade + 2) * (ade + 1)) / 24;
fprintf('Dimensione spazio polinomiale: %d\n', N);

% Inizializzazione matrice degli esponenti in ordine Grlex
chebyshev_indices = zeros(N, 4);
for i = 2:N
    chebyshev_indices(i, :) = mono_next_grlex(4, chebyshev_indices(i - 1, :));
end

% Estrazione coordinate e calcolo della matrice di Vandermonde di riferimento
X = XYZTW_tens_ref(:, 1:4);
V_ref = dCHEBVAND(ade, X, chebyshev_indices); 

% Calcolo dei coefficienti di normalizzazione al quadrato dei polinomi di Chebyshev
[coeffs] = tenscheb_norm2sq(chebyshev_indices);

%**************************************************************************
%
% INIZIO PARTE SHAPE-DEPENDENT
%
%**************************************************************************

% Numero di vertici della mesh tridimensionale
num_vertici = size(vertici_iniziali, 1);

% Inizializzazione dei vertici spazio-temporali
vertici_4D = zeros(2 * num_vertici, 4);

% Configurazione iniziale (tau = 0)
vertici_4D(1:num_vertici, 1:3) = vertici_iniziali;
vertici_4D(1:num_vertici, 4) = 0.0;

% Configurazione finale (tau = 1)
vertici_4D(num_vertici + 1:2*num_vertici, 1:3) = vertici_finali;
vertici_4D(num_vertici + 1:2*num_vertici, 4) = 1.0;


% Calcolo dell'iper-rettangolo che racchiude l'intero dominio spazio-temporale
limiti_min = min(vertici_4D, [], 1);
limiti_max = max(vertici_4D, [], 1);
bbox = [limiti_min; limiti_max];

% Calcolo dei momenti di Chebyshev sul poliedro 4D tramite il teorema della divergenza
moments_ch = chebyshev_moments_polyhedron_4D(vertici_4D, facets, ade, chebyshev_indices, bbox, method);

% Riscalamento della regola di cubatura sulla bounding box reale 4D
XYZTW_tens = scale_rule(XYZTW_tens_ref, bbox);

% Ricombinazione dei pesi di cubatura basata sui momenti del poliedro
W = (XYZTW_tens(:, 5)) .* V_ref * (moments_ch ./ coeffs);
XYZT = XYZTW_tens(:, 1:4);

File: Mesh4D/Fortran4D/Tensor/src/OPC4D_Tensor.f90
MODULE OPC4D_Tensor_Module

   USE TypesDef
   USE ReferenceFunctions
   USE CubatureTensor

   IMPLICIT NONE

CONTAINS

   SUBROUTINE OPC4D_Tensor(ade, vertices_initial, vertices_final, facets, method, XYZT, W)

      !***********************************************************************
      !
      ! SUBROUTINE OPC4D_Tensor(ade, vertices_initial, vertices_final, facets, method, XYZT, W)
      !
      ! Calcola i nodi e i pesi di una regola di cubatura quadridimensionale
      ! su un dominio poliedrico in movimento.
      ! Il dominio spazio-temporale e' il politopo 4D descritto da un poliedro
      ! 3D che si deforma linearmente, per tempo adimensionale tau in [0,1],
      ! dalla configurazione iniziale a quella finale, mantenendo invariata
      ! la connettivita' della mesh superficiale triangolare.
      !
      ! Il metodo non richiede una decomposizione volumetrica del dominio 4D,
      ! i momenti polinomiali vengono ricondotti, mediante il teorema della
      ! divergenza, a integrali sulle ipersuperfici laterali 3D (prismi
      ! triangolari spazio-temporali generati da ciascuna faccia).
      !
      !***********************************************************************
      !
      ! INPUTS:
      !
      !   - ade:
      !       Grado polinomiale totale massimo della regola di cubatura.
      !       Il numero di momenti utilizzati e'
      !       N_mom = (ade+1)(ade+2)(ade+3)(ade+4)/24.
      !
      !   - vertices_initial:
      !       Matrice N x 3 con le coordinate [x_i, y_i, z_i] dei vertici
      !       della mesh superficiale all'istante iniziale (tau = 0).
      !
      !   - vertices_final:
      !       Matrice N x 3 con le coordinate dei medesimi vertici
      !       all'istante finale (tau = 1). Deve avere lo stesso numero di
      !       righe e lo stesso ordinamento di vertices_initial.
      !
      !   - facets:
      !       Matrice M x 3 con la connettivita' della mesh triangolare.
      !       Le facce devono formare una superficie chiusa (ogni spigolo
      !       condiviso da esattamente due facce) e orientata coerentemente
      !       verso l'esterno: l'orientazione determina il segno dei
      !       contributi al teorema della divergenza.
      !
      !   - method:
      !       Metodo di quadratura sul prisma spazio-temporale di ogni faccia:
      !       'DCC'  : Dunavant sul triangolo + Clenshaw-Curtis nel tempo;
      !       'DGL'  : Dunavant sul triangolo + Gauss-Legendre nel tempo;
      !       'GJCC' : Gauss-Jacobi sul triangolo + Clenshaw-Curtis nel tempo;
      !       'GJL'  : Gauss-Jacobi sul triangolo + Gauss-Legendre nel tempo.
      !
      !***********************************************************************
      !
      ! OUTPUTS:
      !
      !   - XYZT:
      !       Matrice n_nodes x 4 con le coordinate [x_k, y_k, z_k, tau_k]
      !       dei nodi di cubatura nel dominio fisico spazio-temporale.
      !
      !   - W:
      !       Vettore n_nodes x 1 con i pesi associati ai nodi.
      !
      !***********************************************************************

      IMPLICIT NONE

      !**********************************************************************
      ! Argomenti
      !**********************************************************************
      INTEGER, INTENT(IN)                         :: ade
      REAL(dp), INTENT(IN)                        :: vertices_initial(:,:)
      REAL(dp), INTENT(IN)                        :: vertices_final(:,:)
      INTEGER, INTENT(IN)                         :: facets(:,:)
      CHARACTER(LEN=*), INTENT(IN)                :: method
      REAL(dp), ALLOCATABLE, INTENT(OUT)          :: XYZT(:,:)
      REAL(dp), ALLOCATABLE, INTENT(OUT)          :: W(:)
      !**********************************************************************
      ! Variabili locali
      !**********************************************************************
      INTEGER                                     :: n_nodes, n_vertices, n_mom, k
      INTEGER                                     :: ind_curr(4)
      REAL(dp), ALLOCATABLE                       :: XYZTW_ref(:,:), XYZTW(:,:), X_ref(:,:)
      REAL(dp), ALLOCATABLE                       :: V_ref(:,:), coeffs(:), moments_ch(:), alpha(:)
      REAL(dp), ALLOCATABLE                       :: vertices_4d(:,:)
      INTEGER, ALLOCATABLE                        :: chebyshev_indices(:,:)
      REAL(dp)                                    :: bbox(2,4)
      !**********************************************************************

      !**********************************************************************
      ! INIZIO PARTE SHAPE-INDEPENDENT
      !**********************************************************************

      ! Griglia tensoriale di Gauss-Chebyshev sull'ipercubo di riferimento [-1,1]^4
      CALL cub_gausscheb_tens4D(2 * ade, XYZTW_ref)
      n_nodes = SIZE(XYZTW_ref, 1)

      ! Dimensione dello spazio polinomiale 4D di grado totale <= ade
      n_mom = ((ade + 1) * (ade + 2) * (ade + 3) * (ade + 4)) / 24

      ! Indici (i,j,k,l) dei polinomi tensoriali di Chebyshev, ordinamento GRLEX
      ALLOCATE(chebyshev_indices(n_mom, 4))
      chebyshev_indices(1,:) = 0
      DO k = 2, n_mom
         ind_curr = chebyshev_indices(k - 1, :)
         CALL mono_next_grlex(4, ind_curr)
         chebyshev_indices(k,:) = ind_curr
      END DO

      ! Matrice di Vandermonde-Chebyshev 4D valutata sui nodi di riferimento
      ALLOCATE(X_ref(n_nodes, 4))
      X_ref = XYZTW_ref(:,1:4)
      CALL dCHEBVAND(ade, X_ref, chebyshev_indices, V_ref)

      ! Norme quadre (pesate) dei polinomi di Chebyshev tensoriali
      CALL tenscheb_norm2sq(chebyshev_indices, coeffs)

      !**********************************************************************
      ! INIZIO PARTE SHAPE-DEPENDENT
      !**********************************************************************

      ! Assemblaggio dei vertici 4D: le prime n_vertices righe sono la
      ! configurazione iniziale (tau = 0), le successive quella finale (tau = 1)
      n_vertices = SIZE(vertices_initial, 1)
      ALLOCATE(vertices_4d(2 * n_vertices, 4))
      vertices_4d(1:n_vertices, 1:3) = vertices_initial
      vertices_4d(1:n_vertices, 4) = 0.0_dp
      vertices_4d(n_vertices+1:2*n_vertices, 1:3) = vertices_final
      vertices_4d(n_vertices+1:2*n_vertices, 4) = 1.0_dp

      ! Bounding box 4D: [xmin,ymin,zmin,tmin ; xmax,ymax,zmax,tmax]
      DO k = 1, 4
         bbox(1,k) = MINVAL(vertices_4d(:,k))
         bbox(2,k) = MAXVAL(vertices_4d(:,k))
      END DO

      ! Momenti della base di Chebyshev sul politopo 4D
      CALL chebyshev_moments_polyhedron_4D(vertices_4d, facets, ade, chebyshev_indices, bbox, method, moments_ch)

      ! Riscalamento dei nodi di riferimento sulla bounding box reale
      CALL scale_rule(XYZTW_ref, bbox, XYZTW)

      !**********************************************************************
      ! Costruzione dei pesi di cubatura
      !**********************************************************************

      ALLOCATE(alpha(n_mom), W(n_nodes), XYZT(n_nodes, 4))

      alpha = moments_ch / coeffs

      W = XYZTW(:,5) * MATMUL(V_ref, alpha)
      
      XYZT = XYZTW(:,1:4)

   END SUBROUTINE OPC4D_Tensor

END MODULE OPC4D_Tensor_Module